helper/winsvc/service_windows.go

```diff
@@ -1,46 +1,85 @@
 // Copyright (c) HashiCorp, Inc.
 // SPDX-License-Identifier: BUSL-1.1
 
-//go:build windows
-// +build windows
-
 package winsvc
 
 import (
-	wsvc "golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/eventlog"
 )
 
-type serviceWindows struct{}
+// Commands that are currently supported
+const SERVICE_ACCEPTED_COMMANDS = svc.AcceptStop | svc.AcceptShutdown
+
+type serviceWindows struct {
+	evtLog Eventlog
+}
 
 func init() {
-	interactive, err := wsvc.IsAnInteractiveSession()
+	isSvc, err := svc.IsWindowsService()
 	if err != nil {
 		panic(err)
 	}
-	// Cannot run as a service when running interactively
-	if interactive {
+	// This should only run when running
+	// as a service
+	if !isSvc {
 		return
 	}
-	go wsvc.Run("", serviceWindows{})
+
+	go executeWindowsService()
 }
 
 // Execute implements the Windows service Handler type. It will be
 // called at the start of the service, and the service will exit
 // once Execute completes.
-func (serviceWindows) Execute(args []string, r <-chan wsvc.ChangeRequest, s chan<- wsvc.Status) (svcSpecificEC bool, exitCode uint32) {
-	const accCommands = wsvc.AcceptStop | wsvc.AcceptShutdown
-	s <- wsvc.Status{State: wsvc.Running, Accepts: accCommands}
+func (srv serviceWindows) Execute(args []string, r <-chan svc.ChangeRequest, s chan<- svc.Status) (svcSpecificEC bool, exitCode uint32) {
+	const accCommands = svc.AcceptStop | svc.AcceptShutdown
+	s <- svc.Status{State: svc.Running, Accepts: SERVICE_ACCEPTED_COMMANDS}
+	srv.evtLog.Info(uint32(EventServiceStarting), "service starting")
+LOOP:
 	for {
-		c := <-r
-		switch c.Cmd {
-		case wsvc.Interrogate:
-			s <- c.CurrentStatus
-		case wsvc.Stop, wsvc.Shutdown:
-			s <- wsvc.Status{State: wsvc.StopPending}
-			chanGraceExit <- 1
-			return false, 0
+		select {
+		case c := <-r:
+			switch c.Cmd {
+			case svc.Interrogate:
+				s <- c.CurrentStatus
+			case svc.Stop, svc.Shutdown:
+				srv.evtLog.Info(uint32(EventLogMessage), "service stop requested")
+				s <- svc.Status{State: svc.StopPending}
```

Comment:
it's tricky to follow some of these channel chains, especially across the separate PRs.

"exit" or "shutdown" channels, in my exp, usually get `close`d to signify that anything (and everything) watching the channel should stop and wrap up what they're doing, and/or throw away the reference to the thing that's been shutdown.  (a context being canceled is this very thing, with its `ctx.Done()` channel). it feels funny to send a single `int` down one.

I think this might be part of why it also feels funny to have the package-level channel in the first place.  the notion of shutting down an entire package seems kind of wild, like it "knows" that the entire program is also shutting down?

I bet this is all fine, but I have some difficulty forming a solid mental model around it.

Reply:
welp, my line-number selection excluded the one line that I spent most of that comment talking about

```go
chanGraceExit <- 1
```